File: vol4/192.py
import math
from fractions import Fraction, gcd

N = 10 ** 5
L = 10 ** 12

def solve(n):
    a = [1, (n + 1)]
    b = [1, 2]
    while b[-1] / gcd(a[-1], b[-1]) <= L:
        a.append(2 * a[-1] + (n - 1) * a[-2])
        b.append(2 * b[-1] + (n - 1) * b[-2])
    f1 = Fraction(a[-2], b[-2])
    f2 = Fraction(a[-3], b[-3])
    mean = (f1 + f2) / 2
    if mean * mean < n:
        f = max(f1, f2)
    else:
        f = min(f1, f2)
    return f.denominator

if __name__ == "__main__":
    ans = 0
    for n in range(2, N + 1):
        print n
        s = int(math.sqrt(n))
        if s * s != n:
            ans += solve(n)
    print ans

File: vol4/192.hs
import Data.List ((\\))
import Data.Ratio ((%), denominator)

sqrt' :: Integer -> Integer
sqrt' = floor . sqrt . fromIntegral

next :: (Integral a) => a -> a -> (a, (a, a)) -> (a, (a, a))
next x s (a, (p, q)) = let p' = a * q - p
                           q' = (x - p' * p') `div` q
                       in ((s + p') `div` q', (p', q'))

fractions :: Integer -> [Integer]
fractions x = map fst $ iterate (next x (sqrt' x)) (sqrt' x, (0, 1))

convergents :: Integer -> [(Integer, Integer)]
convergents x = cs
    where cs = (1, 0) : zipWith3 f (fractions x) cs ((0, 1) : cs)
          f a (n1, d1) (n2, d2) = (a * n1 + n2, a * d1 + d2)

bestDenominator :: Integer -> Integer -> Integer
bestDenominator d x 
    | abs ((n1 % d1) ^ 2 - x') < abs (s ^ 2 - x') = d1
    | otherwise = denominator s
    where x' = fromIntegral x
          ((n1, d1) : (n2, d2) : _) = reverse . takeWhile ((<= d) . snd) $ convergents x
          s = (n1 * ((d - d2) `div` d1) + n2) % (d1 * ((d - d2) `div` d1) + d2)

solve :: Integer -> Integer -> Integer
solve d n = sum . map (bestDenominator d) $ [2..n] \\ (map (^2) [2..sqrt' n])

main :: IO ()
main = print $ solve (10 ^ 12) 100000
